tachyon/zk/air/constraint_system/constraint_system.h

```diff
@@ -0,0 +1,95 @@
+#ifndef TACHYON_ZK_AIR_CONSTRAINT_SYSTEM_CONSTRAINT_SYSTEM_H_
+#define TACHYON_ZK_AIR_CONSTRAINT_SYSTEM_CONSTRAINT_SYSTEM_H_
+
+#include <stddef.h>
+
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "third_party/eigen3/Eigen/Core"
+
+#include "tachyon/math/matrix/matrix_types.h"
+#include "tachyon/zk/air/constraint_system/variable.h"
+#include "tachyon/zk/air/expressions/air_evaluator.h"
+
+namespace tachyon::zk::air {
+
+template <typename F>
+class ConstraintSystem {
+ public:
+  ConstraintSystem() = default;
+  ConstraintSystem(size_t preprocessed_width, size_t main_width,
+                   size_t num_public_values)
+      : preprocessed_width_(preprocessed_width),
+        main_width_(main_width),
+        num_public_values_(num_public_values) {}
+
+  void EnforceFirstRowConstraint(std::unique_ptr<Expression<F>> constraint) {
+    EnforceConstraint(ExpressionFactory<F>::FirstRow(std::move(constraint)));
+  }
+
+  void EnforceLastRowConstraint(std::unique_ptr<Expression<F>> constraint) {
+    EnforceConstraint(ExpressionFactory<F>::LastRow(std::move(constraint)));
+  }
+
+  void EnforceTransitionConstraint(std::unique_ptr<Expression<F>> constraint) {
+    EnforceConstraint(ExpressionFactory<F>::Transition(std::move(constraint)));
+  }
+
+  void EnforceConstraint(std::unique_ptr<Expression<F>> constraint) {
+    constraints_.push_back(std::move(constraint));
+  }
+
+  bool IsSatisfied(AirEvaluator<F>& evaluator,
+                   const std::vector<F>& public_values,
+                   std::unique_ptr<const math::RowMajorMatrix<F>> preprocessed,
+                   const math::RowMajorMatrix<F>& main) const {
```

Comment:
Please do not transfer ownership since this function doesn't own the matrix in the end. Just pass it as a pointer. Also, please place the optional argument at the end.

```suggestion
                   const math::RowMajorMatrix<F>& main
                   const math::RowMajorMatrix<F>* preprocessed = nullptr) const {
```

Reply:
Also, I found that your code prioritizes `preprocessed` over `main`, but considering that `preprocessed` is optional, the order should be the other way around.